ArduCopter/mode_autorotate.cpp

```diff
@@ -111,145 +164,194 @@ void ModeAutorotate::run()
+        // Attitude controller is updated in navigation switch-case statements
+        g2.arot.calc_avg_acc_z();
 
-                // Prevent running the initial glide functions again
-                _flags.ss_glide_initial = false;
-            }
+        break;
+    }
 
-            // Run airspeed/attitude controller
-            g2.arot.set_dt(G_Dt);
-            g2.arot.update_forward_speed_controller();
+    case Autorotation_Phase::FLARE: {
+        // Steady state glide functions to be run only once
+        if (_flags.flare_initial == true) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Flare_Phase");
 
-            // Retrieve pitch target 
-            _pitch_target = g2.arot.get_pitch();
+            // Set following trim low pass cut off frequency
+            g2.arot.set_col_cutoff_freq(g2.arot.get_col_glide_freq());
+
+            // Set target head speed in head speed controller
+            _target_head_speed = HEAD_SPEED_TARGET_RATIO;  //Ensure target hs is set to glide incase hs has not reached target for glide
+            g2.arot.set_target_head_speed(_target_head_speed);
+            g2.arot.get_entry_speed();
+            // Prevent running the initial flare functions again
+            _flags.flare_initial = false;
+        }
+        // Run flare controller
+        g2.arot.flare_controller();
 
-            // Update head speed/ collective controller
-            _flags.bad_rpm = g2.arot.update_hs_glide_controller(G_Dt); 
-            // Attitude controller is updated in navigation switch-case statements
+        //update sink rate derivative
+        g2.arot.calc_sink_d_avg();
 
-            break;
+        // Update head speed/ collective controller
+        _flags.bad_rpm = g2.arot.update_hs_glide_controller();
+        // Retrieve pitch target
+        _pitch_target = g2.arot.get_pitch();
+
+        //calc average acceleration on z axis for estimating flare effectiveness
+        g2.arot.calc_avg_acc_z();
+
+        break;
+    }
+    case Autorotation_Phase::TOUCH_DOWN: {
+        if (_flags.touch_down_initial == true) {
+            gcs().send_text(MAV_SEVERITY_INFO, "Touchdown_Phase");
+            // Prevent running the initial touchdown functions again
+            _flags.touch_down_initial = false;
+            _touchdown_time_ms = millis();
+            g2.arot.set_col_cutoff_freq(g2.arot.get_col_cushion_freq());
+            g2.arot.set_entry_sink_rate(curr_vel_z);
+            g2.arot.set_entry_alt(g2.arot.get_ground_distance());
+            g2.arot.set_ground_clearance(copter.rangefinder.ground_clearance_cm_orient(ROTATION_PITCH_270));
         }
+        g2.arot.touchdown_controller();
+        _pitch_target = g2.arot.get_pitch();
 
-        case Autorotation_Phase::FLARE:
-        case Autorotation_Phase::TOUCH_DOWN:
-        {
-            break;
+        if (fabsf(inertial_nav.get_velocity_z_up_cms()) < 10) {
+            copter.ap.land_complete = true;
```

Comment:
```suggestion
            copter.set_land_complete(true);
```
This does make me a bit nervous. This is the only caller outside the landing detector that sets land complete true, other modes will set it false. This is effectively short circuiting the landing detector. This also means the landing detection logic for this mode lives here not with all the other landing detection code in `land_detector.cpp` which means if that code it updated its likely this will be missed and get broken. 

What is the need for faster landing detection? I'm not sure what we do with throttle if the motor came back to life would it start sooling up again?

Reply:
We have already disengaged the interlock to get into the autorotation mode.  So we are not in any risk of motors spooling up again.